### PR TITLE
[P3 tests][Team3] Add Task3: Compressor Correctness Tests

### PR DESCRIPTION
**Pull request by ZTShao**
What's your team number?
Team 3

What is the functionality being tested?
Task3: Compressor Correctness

How many test cases do you have?
8

Describe your tests briefly:
testcase1 deals with the easiest encoding with input 123456. It aims to test the delta encoding which will turn it into 1 1 1 1 1 1.

testcase2 deals with input that will be compressed using multiple bytes.

testcase3 tests whether the encoder could handle the Integer.MAX_VALUE.

testcases 4-6 are just inverse tests of 1-3, which test the decoding of them.

testcases 7,8 test decoding from the chosen offset and length.

### CODE REVIEW

File: src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        byte[] expected3 = {(byte)0x87,(byte)0xff,(byte)0xff,(byte)0xff,(byte)0x7f};
+        assertArrayEquals(expected3,deltaVarLenCompressor.encode(integers3));
+    }
+    //Test decoding of all bytes.
+    @Test
+    public void test4() {
+        List<Integer> expected1 = Arrays.asList(1, 2, 3, 4, 5, 6); //after delta encoding 1,1,1,1,1,1
+        byte[] bytes1 = {(byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x01};
+        List<Integer> actual1 = deltaVarLenCompressor.decode(bytes1);
+        assertEquals(actual1.size(),expected1.size());
```
Comment (LaytonW) [inline src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java]: Expected value should be the 1st argument in `assertEquals`
```suggestion
        assertEquals(expected1.size(), actual1.size());
```

File: src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        }
+    }
+
+    @Test
+    public void test5() {
+        List<Integer> expected2 = Arrays.asList(128, 384, 896, 1920, 3968);//after delta encoding 128,256,512,1024,2048
+        byte[] bytes2 = {(byte) 0x81, (byte) 0x00, (byte) 0x82, (byte) 0x00, (byte) 0x84, (byte) 0x00, (byte) 0x88, (byte) 0x00,
+                (byte) 0x90, (byte) 0x00};
+        List<Integer> actual2 = deltaVarLenCompressor.decode(bytes2);
+        assertEquals(actual2.size(),expected2.size());
```
Comment (LaytonW) [inline src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java]:
```suggestion
        assertEquals(expected2.size(), actual2.size());
```

File: src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+        for(int i=0;i<expected2.size();i++){
+            assertEquals(expected2.get(i),actual2.get(i));
+        }
+    }
+    @Test
+    public void test6() {
+        List<Integer> expected3 = Arrays.asList(Integer.MAX_VALUE);
+        byte[] bytes3 = {(byte)0x87,(byte)0xff,(byte)0xff,(byte)0xff,(byte)0x7f};
+        List<Integer> actual3 = deltaVarLenCompressor.decode(bytes3);
+        assertEquals(actual3.size(),expected3.size());
```
Comment (LaytonW) [inline src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java]:
```suggestion
        assertEquals(expected3.size(), actual3.size());
```

File: src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        }
+    }
+    //Test decoding of some of the bytes
+    @Test
+    public void test7() {
+        List<Integer> original = Arrays.asList(1, 2, 3, 4, 5, 6); //after delta encoding 1,1,1,1,1,1
+        byte[] bytes1 = {(byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x01, (byte) 0x01};
+        List<Integer> actual1 = deltaVarLenCompressor.decode(bytes1,4,2);
+        List<Integer> expected1 = Arrays.asList(5,6);
+        assertEquals(actual1.size(),expected1.size());
```
Comment (LaytonW) [inline src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java]:
```suggestion
        assertEquals(expected1.size(), actual1.size());
```

File: src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Test
+    public void test8() {
+        List<Integer> original = Arrays.asList(128, 384, 896, 1920, 3968);//after delta encoding 128,256,512,1024,2048
+        byte[] bytes2 = {(byte) 0x81, (byte) 0x00, (byte) 0x82, (byte) 0x00, (byte) 0x84, (byte) 0x00, (byte) 0x88, (byte) 0x00,
+                (byte) 0x90, (byte) 0x00};
+        List<Integer> actual2 = deltaVarLenCompressor.decode(bytes2,4,6);
+        List<Integer> expected2 = Arrays.asList(896,1920,3968);
+        assertEquals(actual2.size(),expected2.size());
```
Comment (LaytonW) [inline src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java]:
```suggestion
        assertEquals(expected2.size(), actual2.size());
```

Comment (ZTShao) [reply src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java]: Yes, it seems I misunderstand the aim of decode function, thanks!

Comment (ZTShao) [reply src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java]: I have modified them. Thanks!

**Comment by zhuangliulz** [comment]
### Review from Team 2

A solid test covering most potential scenarios. Since the major problems such as in `test7` and `test8` had been addressed, we suggest some minor modifications. More details can be seen in single comments.

File: src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+
+public class Team3CompressorTest {
+    private DeltaVarLenCompressor deltaVarLenCompressor;
+    private NaiveCompressor naiveCompressor;
```
Comment (zhuangliulz) [inline src/test/java/edu/uci/ics/cs221/index/positional/Team3CompressorTest.java]: redundant code

**Pull request merged**